File: scripts/ingest_docs.js
import { createClient } from '@supabase/supabase-js';
import { GoogleGenerativeAI } from '@google/generative-ai';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';
const require = createRequire(import.meta.url);
const pdfParse = require('pdf-parse');
import officeparser from 'officeparser';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const supabaseUrl = process.env.VITE_SUPABASE_URL || process.env.SUPABASE_URL;
const supabaseKey = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.VITE_SUPABASE_ANON_KEY;
const geminiApiKey = process.env.VITE_GEMINI_API_KEY || process.env.GEMINI_API_KEY;

if (!supabaseUrl || !supabaseKey || !geminiApiKey) {
  console.error("Error: Missing required environment variables.");
  process.exit(1);
}

const supabase = createClient(supabaseUrl, supabaseKey);
const genAI = new GoogleGenerativeAI(geminiApiKey);
const model = genAI.getGenerativeModel({ model: "gemini-embedding-001" });

function chunkText(text, chunkSize = 1000, overlap = 100) {
  const chunks = [];
  let start = 0;
  while (start < text.length) {
    let end = start + chunkSize;
    chunks.push(text.slice(start, end));
    start += chunkSize - overlap;
  }
  return chunks;
}

async function processFile(filepath) {
  console.log(`Processing ${filepath}...`);
  const filename = path.basename(filepath);
  const ext = path.extname(filepath).toLowerCase();
  
  try {
    let text = '';
    
    if (ext === '.pdf') {
        const dataBuffer = fs.readFileSync(filepath);
        const data = await pdfParse(dataBuffer);
        text = data.text;
    } else if (ext === '.pptx' || ext === '.docx') {
        const raw = await officeparser.parseOffice(filepath);
        text = raw.toText ? raw.toText() : String(raw);
    } else if (ext === '.txt' || ext === '.md' || ext === '.csv') {
        text = fs.readFileSync(filepath, 'utf8');
    } else {
        console.log(`  Skipping unsupported file type: ${ext}`);
        return;
    }

    const chunks = chunkText(text);

    for (let i = 0; i < chunks.length; i++) {
        const chunk = chunks[i];
        if (!chunk.trim()) continue;

        console.log(`  Embedding chunk ${i + 1}/${chunks.length}...`);
        
        try {
            const result = await model.embedContent(chunk);
            const embedding = result.embedding.values;

            const { error } = await supabase.from('documents').insert({
                content: chunk,
                metadata: { source: filename, chunk_index: i },
                embedding: embedding
            });

            if (error) {
                console.error(`  Error inserting chunk ${i}:`, error.message);
            }
        } catch (embedError) {
             console.error(`  Error generating embedding for chunk ${i}:`, embedError.message);
        }
    }
  } catch (err) {
      console.error(`Error reading file ${filepath}:`, err.message);
  }
}

async function main() {
  const docsDir = path.join(__dirname, '../data/docs');
  if (!fs.existsSync(docsDir)) {
    console.error(`Directory not found: ${docsDir}`);
    return;
  }

  console.log("Connecting to Supabase and processing documents...");
  const files = fs.readdirSync(docsDir);

  for (const file of files) {
    const ext = path.extname(file).toLowerCase();
    const validExts = ['.txt', '.md', '.csv', '.pdf', '.pptx', '.docx'];
    if (validExts.includes(ext)) {
      const filepath = path.join(docsDir, file);
      await processFile(filepath);
    }
  }
  
  console.log("Document ingestion complete!");
}

main();

File: scripts/ingest_docs.py
import os
import json
import httpx
from supabase import create_client, Client

# Environment Variables
SUPABASE_URL = os.environ.get("SUPABASE_URL")
SUPABASE_KEY = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")

if not SUPABASE_URL or not SUPABASE_KEY or not GEMINI_API_KEY:
    print("Error: Missing SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY, or GEMINI_API_KEY environment variables.")
    exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

def get_gemini_embedding(text):
    url = f"https://generativelanguage.googleapis.com/v1beta/models/text-embedding-004:embedContent?key={GEMINI_API_KEY}"
    headers = {'Content-Type': 'application/json'}
    data = {
        "model": "models/text-embedding-004",
        "content": {
            "parts": [{
                "text": text
            }]
        }
    }
    
    response = httpx.post(url, headers=headers, json=data, timeout=30.0)
    response.raise_for_status()
    result = response.json()
    return result['embedding']['values']

def chunk_text(text, chunk_size=1000, overlap=100):
    chunks = []
    start = 0
    while start < len(text):
        end = start + chunk_size
        chunk = text[start:end]
        chunks.append(chunk)
        start += chunk_size - overlap
    return chunks

def process_file(filepath):
    print(f"Processing {filepath}...")
    with open(filepath, 'r', encoding='utf-8') as f:
        text = f.read()
        
    filename = os.path.basename(filepath)
    chunks = chunk_text(text)
    
    for i, chunk in enumerate(chunks):
        if not chunk.strip():
            continue
            
        print(f"  Embedding chunk {i + 1}/{len(chunks)}...")
        try:
            embedding = get_gemini_embedding(chunk)
            
            # Store in Supabase
            data, count = supabase.table('documents').insert({
                "content": chunk,
                "metadata": {"source": filename, "chunk_index": i},
                "embedding": embedding
            }).execute()
            
        except Exception as e:
            print(f"  Error processing chunk {i}: {e}")

def main():
    docs_dir = "../data/docs"
    if not os.path.exists(docs_dir):
        print(f"Directory not found: {docs_dir}")
        return
        
    print("Connecting to Supabase and processing documents...")
    for filename in os.listdir(docs_dir):
        if filename.endswith(".txt"):
            filepath = os.path.join(docs_dir, filename)
            process_file(filepath)
            
    print("Document ingestion complete!")

if __name__ == "__main__":
    main()
